spoc/pixels.py

```diff
@@ -31,291 +98,105 @@ class GenomicBinner:
-                "end_2",
-                "chrom_3",
-                "start_3",
-                "end_3",
-            ],
-        ]
-        sister_triplets_cis_trans_cis = contacts.loc[
-            is_cis_trans_cis,
-            [
-                "read_name",
-                "chrom_1",
-                "start_1",
-                "end_1",
-                "chrom_3",
-                "start_3",
-                "end_3",
-                "chrom_2",
-                "start_2",
-                "end_2",
-            ],
-        ]
-        sister_triplets_trans_cis_cis = contacts.loc[
-            is_trans_cis_cis,
-            [
-                "read_name",
-                "chrom_2",
-                "start_2",
-                "end_2",
-                "chrom_3",
-                "start_3",
-                "end_3",
-                "chrom_1",
-                "start_1",
-                "end_1",
-            ],
-        ]
-        # concatenate and return
-        return dd.concat(
-            list(
-                map(
-                    self._reset_contact_columns,
-                    [
-                        sister_triplets_cis_cis_trans,
-                        sister_triplets_cis_trans_cis,
-                        sister_triplets_trans_cis_cis,
-                    ],
-                )
-            )
-        )
-
-    def _flip_symmetric_contacts(self, contacts: dd.DataFrame) -> dd.DataFrame:
-        """Flips contacts to upper triangular such that
-        start_1 <= start_2. This is a Triplet specific implementation.
-        TODO: Think about what this means for other contact order.
-        Symmetry is not trivial for higher dimensional matrices (see
-        https://math.stackexchange.com/questions/615119/how-to-generalize-symmetry-for-higher-dimensional-arrays).
-        Ideally, we would have a contact class that has a notion of symmetry, and flipping would be offloaded
-        to that contact class. For example, non-sister sensitive triplets would have full symmetry, i.e. permutations
-        of indices do not matter. However, sister-sensitive triplets have a more restricted notion of symmetry since
-        only permutation of the first two indices leaves information unchanged. This method implements contact flipping
-        based on the latter notion of symmetry.
-        """
-        # create boolean indexer
-        is_lower_triangular = (contacts.chrom_1 == contacts.chrom_2) & (
-            contacts.start_1 > contacts.start_2
-        )
-        # select and flip
-        lower_flipped = contacts.loc[is_lower_triangular, :].rename(
-            columns={
-                "start_1": "start_2",
-                "start_2": "start_1",
-                "end_1": "end_2",
-                "end_2": "end_1",
+            return self._get_assigned_bin_output_structure()
+        return data_frame.assign(
+            **{
+                f"start_{index}": (data_frame[f"pos_{index}"] // self._bin_size) * self._bin_size
```

Comment:
This works for now with fixed bin size, but for the future's sake we should also be able to use arbitrary bins (as a dataframe). This can be useful for SPRITE (e.g. annotating RNAs) or equal bins with offsets (e.g. 100Kb bins with offset of 50Kb: 50kb, 150kb, 250kb, ...) It will reduce performance, though, so fixed bins and arbitrary bins should be implemented separately.

Reply:
Good point, added it as a requirement to this ticket: https://trello.com/c/93ffI2Hn/78-add-variable-binsize-handling